File: ios/YuQiuPu/DataSync.swift
import Foundation

enum DataSync {
    private static let remoteURL = URL(string: "https://your-bucket.oss-cn-hangzhou.aliyuncs.com/yuqiupu.db")!

    /// Download the latest DB from the remote URL and replace the local copy.
    /// Failures are silently ignored — the app continues with local data.
    static func syncIfNeeded() {
        let task = URLSession.shared.downloadTask(with: remoteURL) { tempURL, response, error in
            guard let tempURL = tempURL, error == nil else {
                print("[DataSync] Download failed: \(error?.localizedDescription ?? "unknown error")")
                return
            }

            // Verify we got a successful HTTP response
            if let httpResponse = response as? HTTPURLResponse,
               !(200...299).contains(httpResponse.statusCode) {
                print("[DataSync] Server returned status \(httpResponse.statusCode)")
                return
            }

            // Move to a stable temporary location (the download temp file may be deleted)
            let stableTemp = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString + ".db")
            do {
                try FileManager.default.moveItem(at: tempURL, to: stableTemp)
            } catch {
                print("[DataSync] Failed to stage downloaded file: \(error)")
                return
            }

            DispatchQueue.main.async {
                Database.shared.replaceWith(downloadedDBAt: stableTemp)
            }
        }
        task.resume()
    }
}

File: android/app/src/main/java/com/yuqiupu/app/data/DataSync.kt
package com.yuqiupu.app.data

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

object DataSync {

    private const val REMOTE_URL =
        "https://your-bucket.oss-cn-hangzhou.aliyuncs.com/yuqiupu.db"

    /**
     * Download the latest DB from the remote URL and replace the local copy.
     * Failures are silently ignored — the app continues with local data.
     */
    suspend fun syncIfNeeded(context: Context) {
        withContext(Dispatchers.IO) {
            try {
                val url = URL(REMOTE_URL)
                val connection = url.openConnection() as HttpURLConnection
                connection.connectTimeout = 10_000
                connection.readTimeout = 30_000

                try {
                    if (connection.responseCode !in 200..299) {
                        return@withContext
                    }

                    val tempFile = File(context.cacheDir, "yuqiupu_download.db")
                    connection.inputStream.use { input ->
                        tempFile.outputStream().use { output ->
                            input.copyTo(output)
                        }
                    }

                    Database.getInstance(context).replaceWith(tempFile)
                } finally {
                    connection.disconnect()
                }
            } catch (e: Exception) {
                // Silently ignore — continue with local data
                e.printStackTrace()
            }
        }
    }
}
